test/pylib/db/writer.py

```diff
@@ -0,0 +1,154 @@
+#
+# Copyright (C) 2022-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+import asyncio
+import sqlite3
+from typing import List
+
+from attr import AttrsInstance, asdict
+
+TESTS_TABLE = 'tests'
+METRICS_TABLE = 'test_metrics'
+SYSTEM_RESOURCE_METRICS_TABLE = 'system_resource_metrics'
+CGROUP_MEMORY_METRICS_TABLE = 'cgroup_memory_metrics'
+DEFAULT_DB_NAME = 'sqlite.db'
+DATE_TIME_TEMPLATE = '%Y-%m-%d %H:%M:%S.%f'
+
+create_table = [
+    f'''
+    CREATE TABLE IF NOT EXISTS {TESTS_TABLE} (
+        id INTEGER PRIMARY KEY,
+        architecture VARCHAR(15) NOT NULL,
+        directory VARCHAR(255),
+        mode VARCHAR(15) NOT NULL,
+        run_id INTEGER,
+        test_name VARCHAR(255) NOT NULL
+    );
+    ''',
+
+    f'''
+    CREATE TABLE IF NOT EXISTS {METRICS_TABLE} (
+        id INTEGER PRIMARY KEY,
+        test_id INT NOT NULL,
+        user_sec REAL,
+        system_sec REAL,
+        usage_sec REAL,
+        memory_peak INTEGER,
+        time_taken REAL,
+        time_start DATETIME,
+        time_end DATETIME,
+        success BOOLEAN,
+        FOREIGN KEY(test_id) REFERENCES {TESTS_TABLE}(id)
+    );
+    ''',
+
+    f'''
+    CREATE TABLE IF NOT EXISTS {SYSTEM_RESOURCE_METRICS_TABLE} (
+        id INTEGER PRIMARY KEY,
+        memory REAL,
+        cpu REAL,
+        timestamp DATETIME
+    );
+    ''',
+
+    f'''
+    CREATE TABLE IF NOT EXISTS {CGROUP_MEMORY_METRICS_TABLE} (
+        id INTEGER PRIMARY KEY,
+        test_id INT NOT NULL,
+        memory REAL,
+        timestamp DATETIME,
+        FOREIGN KEY(test_id) REFERENCES {TESTS_TABLE}(id)
+    );
+    '''
+]
+
+
+class SingletonMeta(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            instance = super().__call__(*args, **kwargs)
+            cls._instances[cls] = instance
+        return cls._instances[cls]
+
+
+class SQLiteWriter(metaclass=SingletonMeta):
+    __instance = None
+
+    def __init__(self, database_path):
+        """
+        Initializes the SQLWriter object.
+
+        Args:
+            database_path: Path to the SQLite database file.
+        """
+        self.lock = asyncio.Lock()
+        self.conn = sqlite3.connect(database_path)
+        self.cursor = self.conn.cursor()
+        self.cursor.execute('PRAGMA foreign_keys=ON')
```

Comment:
let's also issue 'PRAGMA sychronous=off' to disable fsync; fsync can stall other coroutines.